File: python-scripts/internal_distances.py
"""This Python script computes the mean-square internal distances
   for a system of ring polymers."""

import sys
import math
import glob
import pypolymer
import numpy as np

# show result in Matplotlib
mpl = False

# generation of file names: (1) glob, (2) loop, (3) manual or (4) log scale
blockfiles_method = 2

if(blockfiles_method == 1):

  prefix = 'rings400a.'
  existingfiles = glob.glob(prefix + '*')
  existingfiles = filter(lambda u: u[u.rindex('.') + 1:].isdigit(), existingfiles)
  existingfiles.sort(lambda u, v: cmp(int(u[u.rindex('.') + 1:]), int(v[v.rindex('.') + 1:])))
  blockfiles = existingfiles[::]
  if 0:
    maxfiles = 30
    blockfiles = blockfiles[::int(len(blockfiles) / float(maxfiles))] = blockfiles[::]

elif(blockfiles_method == 2):

  istr = 2000000
  iend = 2000000
  incr = 2000000

  blockfiles = []
  for i in range(istr, iend + incr, incr):
    blockfiles.append('../sqt/stripped/' + str(i) + '.pos')

elif(blockfiles_method == 3):

  blockfiles = ['RGEQ_N200M400.end']

elif(blockfiles_method == 4):

  # prefix
  prefix = '../../rings200_26.'

  # number of files per decade
  n = 5

  # exponent of first and last decade (time steps)
  start_exp = 5
  end_exp = 10
 
  # base
  b = math.e

  number_of_decades = end_exp - start_exp
  start = math.log(10**start_exp)
  end = math.log(10**end_exp)
  step = (end - start) / (n * number_of_decades - 1)
  uniform_time = []
  for i in range(n * number_of_decades):
    uniform_time.append(b**(start + step * i))
  existingfiles = glob.glob(prefix + '*')
  existingfiles = filter(lambda u: u[u.index('.', 5) + 1:].isdigit(), existingfiles)
  existingfiles = [int(u[u.index('.', 5) + 1:]) for u in existingfiles]
  existingfiles.sort()
  uniform_time = [pypolymer.find_closest_existing_file(existingfiles, u) for u in uniform_time]
  uniform_time = pypolymer.unique(uniform_time)
  blockfiles = [prefix + str(u) for u in uniform_time]

else:

  sys.stderr.write('ERROR: blockfiles_method\n')
  sys.exit()

sys.stdout.write('\n\n')
sys.stdout.write('=================================================\n')
sys.stdout.write('                      [d(s)]^2                   \n')
sys.stdout.write('=================================================\n')
sys.stdout.write('\n\nTotal number of files: ' + str(len(blockfiles)) + '\n')
if(len(blockfiles) > 2):
  sys.stdout.write('  blockfiles[0] = ' + blockfiles[0] + '\n')
  sys.stdout.write('  blockfiles[1] = ' + blockfiles[1] + '\n')
  sys.stdout.write('  blockfiles[2] = ' + blockfiles[2] + '\n')
  sys.stdout.write('  ...' + '\n')
  sys.stdout.write('  blockfiles[-3] = ' + blockfiles[-3] + '\n')
  sys.stdout.write('  blockfiles[-2] = ' + blockfiles[-2] + '\n')
  sys.stdout.write('  blockfiles[-1] = ' + blockfiles[-1] + '\n\n')

for file in blockfiles:

  fmanager = pypolymer.read_blockfile(file)
  timestep = fmanager[0]
  total_particles = fmanager[1]
  M_rings = fmanager[2]; M_linear = fmanager[3]
  monomers = fmanager[4]; monomers_half = fmanager[5]
  x = fmanager[6]; y = fmanager[7]; z = fmanager[8]

  if(file == blockfiles[0]):
    total_particles1st = total_particles
    sys.stdout.write('System parameters:\n')
    sys.stdout.write('  M_rings = ' + str(M_rings) + '\n')
    sys.stdout.write('  M_linear = ' + str(M_linear) + '\n')
    sys.stdout.write('  N = ' + str(monomers) + '\n')
    sys.stdout.write('  N/2 = ' + str(monomers_half) + '\n\n')
    dsq = np.zeros(monomers_half + 1, dtype=np.float64)
    dsq_dsqave = np.zeros(monomers_half + 1, dtype=np.float64)
    ct = np.zeros(monomers_half + 1, dtype=np.int32)
  if(total_particles1st != total_particles):
    sys.stderr.write('Change in total_particles (' + str(total_particles) + ').\n')
    sys.exit()

  sys.stdout.write('Working on ' + file + '\n')

  for i in range(M_rings):
    for j in range(monomers - 1):
      p1 = i * monomers + j
      xp1 = x[p1]
      yp1 = y[p1]
      zp1 = z[p1]
      for k in range(j + 1, monomers):
        p2 = i * monomers + k
        rsq = (xp1 - x[p2])**2 + (yp1 - y[p2])**2 + (zp1 - z[p2])**2
        s = k - j
        if(s > monomers_half): s = monomers - s
        dsq[s] = dsq[s] + rsq
        ct[s] = ct[s] + 1

dsq = dsq / ct
for s in range(1, monomers_half + 1):
  dsq_dsqave[s] = dsq[s] * (1.0 / s + 1.0 / (monomers - s))

outfile = 'dsq' + str(monomers) + '.dat'
f = open(outfile, 'w')
f.write('# s  dsq(s)/sigma^2  dsq(s)/<dsq>\n')
for s in range(1, monomers_half + 1):
  f.write('%d %.3f %.3f\n' % (s, dsq[s], dsq_dsqave[s]))
f.close()
sys.stdout.write(outfile + ' has been written to disk.\n')

if mpl:
  import matplotlib.pyplot as plt
  plt.semilogx(dsqdsq_ave, 'b-')
  plt.xlim(1, 1000)
  plt.ylim(0.1, 3)
  plt.show()

File: python-scripts/internal_distances.f90
! This F90 code computes the internal distances of polymer
! chains.

! Large integers are needed so compile with the following:
! gfortran -fdefault-integer-8 -O3 internal_distances.f90

program internal_distances

implicit none

integer*4 chains, monomers, monomers_half, total_particles
parameter (chains = 260)
parameter (monomers = 200)
parameter (monomers_half = monomers / 2)
parameter (total_particles = chains * monomers)
integer*4 id(total_particles)
real*8 x(total_particles), y(total_particles), z(total_particles)
real*8 x_buf(total_particles), y_buf(total_particles), z_buf(total_particles)
real*8 xp1, yp1, zp1, rsq
real*8 dsq(monomers_half), dsq_dsqave(monomers_half)
integer*8 ct(monomers_half)

integer*4 i, j, k, p1, p2, s

integer*8 f, istr, iend, incr
character(10) chrf10
character(80) prefix
character(80) flnm

character(80) chrf80
integer*8 timestep
integer*4 np
real*8 xmin, ymin, zmin
real*8 xmax, ymax, zmax
real*8 dx, dy, dz
real*8 rho, drho

ct = 0
dsq = 0.0D0

prefix = '../../rings10_250.'
istr = 1000000
iend = 2000000000
incr = 1000000

do f = istr, iend, incr
 
  write(chrf10, '(i10)') f
  flnm = trim(adjustl(prefix)) // trim(adjustl(chrf10))
  write(*,*) 'Working on file ', flnm

  open(10, file = flnm)
  read(10,*) chrf80
  read(10,*) timestep
  read(10,*) chrf80
  read(10,*) np
  read(10,*) chrf80
  read(10,*) xmin, xmax
  read(10,*) ymin, ymax
  read(10,*) zmin, zmax
  read(10,*) chrf80
  do i = 1, total_particles
    read(10,*) id(i), x_buf(i), y_buf(i), z_buf(i)
  enddo
  close(10)

  ! consistency checks
  if(timestep.ne.f) write(*,*) 'WARNING: Mismatch between timestep and file name: ', timestep, f
  if(np.ne.total_particles) write(*,*) 'WARNING: Mismatch between np and total_particles.'
  rho = np / ((xmax - xmin) * (ymax - ymin) * (zmax - zmin))
  drho = abs(1.0D0 - rho / 0.85D0)
  if(drho.gt.0.01D0) write(*,*) 'WARNING: Mismatch between densities: ', drho

  ! sort by id
  do i = 1, total_particles
    j = id(i)
    x(j) = x_buf(i)
    y(j) = y_buf(i)
    z(j) = z_buf(i)
  enddo

  ! loop over only first 10 chains (which are rings)
  do i = 1, 10
    do j = 1, monomers - 1
      p1 = (i - 1) * monomers + j
      xp1 = x(p1)
      yp1 = y(p1)
      zp1 = z(p1)
      do k = j + 1, monomers
        p2 = (i - 1) * monomers + k
        rsq = (xp1 - x(p2))**2 + (yp1 - y(p2))**2 + (zp1 - z(p2))**2
        s = k - j
        if(s.gt.monomers_half) s = monomers - s
        dsq(s) = dsq(s) + rsq
        ct(s) = ct(s) + 1
      enddo
    enddo
  enddo

enddo

do i = 1, monomers_half
  dsq(i) = dsq(i) / ct(i)
  dsq_dsqave(i) = dsq(i) * (1.0 / i + 1.0 / (monomers - i))
enddo

open(10, file = 'dsq.dat')
write(10,*) '# s  dsq(s)/sigma^2  dsq(s)/<dsq>'
do i = 1, monomers_half
  write(10,'(i5,f8.3,f7.3)') i, dsq(i), dsq_dsqave(i)
enddo
close(10)

end program
